core/src/main/java/hudson/model/Slave.java

```diff
@@ -412,13 +417,44 @@ public URL getURL() throws IOException {
      *      If there is no computer it will return a {@link hudson.Launcher.DummyLauncher}, otherwise it
      *      will return a {@link hudson.Launcher.RemoteLauncher} instead.
      */
+    @Nonnull
     public Launcher createLauncher(TaskListener listener) {
         SlaveComputer c = getComputer();
         if (c == null) {
-            listener.error("Issue with creating launcher for agent " + name + ".");
+            listener.error("Issue with creating launcher for agent " + name + ". Computer has been disconnected");
             return new Launcher.DummyLauncher(listener);
         } else {
-            return new RemoteLauncher(listener, c.getChannel(), c.isUnix()).decorateFor(this);
+            // TODO: ideally all the logic below should be inside the SlaveComputer class with proper locking to prevent race conditions, 
+            // but so far there is no locks for setNode() hence it requires serious refactoring
+            
+            // Ensure that the Computer instance still points to this node
+            // Otherwise we may end up running the command on a wrong (reconnected) Node instance.
+            Slave node = c.getNode();
+            if (node != this) {
+                String message = "Issue with creating launcher for agent " + name + ". Computer has been reconnected";
+                if (LOGGER.isLoggable(Level.WARNING)) {
+                    LOGGER.log(Level.WARNING, message, new IllegalStateException("Computer has been reconnected, this Node instance cannot be used anymore"));
+                }
+                return new Launcher.DummyLauncher(listener);
+            }
+            
+            // RemoteLauncher requires an Channel instance to operate correctly
+            final Channel channel = c.getChannel();
+            final Boolean isUnix = channel != null ? c.isUnix() : null; // No sense to check it if the channel is down
```

Comment:
I expected `channel.call()` to handle it, but it does not...
The only handler is https://github.com/jenkinsci/remoting/blob/a762e0703a04ae41fbd413b2379d57f80fc282b2/src/main/java/hudson/remoting/Channel.java#L604-L605 , which effectively checks for the closed channel after a bunch of operations.

IMHO `Channel` should provide a new API for user-space calls which should not be performed in parallel with close. Making it a default behavior seems to be a breaking change though, needs much testing